### ISSUE TITLE
setup.py does not set sys.path correctly for importing dbusmock (when using PEP517 backend)

### ISSUE THREAD

**Issue by mgorny** [martinpitt/python-dbusmock#142]
When using PEP517 backend, there is no guarantee that `.` will be present on `sys.path`. This causes e.g. the following build failure (or even worse, silently using version from installed dbusmock):

```pytb
$ gpep517 build-wheel --output-fd 1 --wheel-dir /tmp
Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.11/gpep517", line 4, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/lib/python3.11/site-packages/gpep517/__main__.py", line 191, in main
    return func(args)
           ^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/gpep517/__main__.py", line 87, in build_wheel
    wheel_name = backend.build_wheel(args.wheel_dir, args.config_json)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 244, in build_wheel
    return self._build_with_temp_dir(['bdist_wheel'], '.whl',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 229, in _build_with_temp_dir
    self.run_setup()
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(code, locals())
  File "<string>", line 9, in <module>
ModuleNotFoundError: No module named 'dbusmock'
```

Will submit a PR shortly.

**Comment by martinpitt** [comment]
Right, I [noticed something similar](https://github.com/martinpitt/python-dbusmock/runs/7377001530?check_suite_focus=true) in CI yesterday. I considered this a regression in Debian unstable/Ubuntu devel series. setuptools did not need this `sys.path` wrangling for many years. I attributed this to the current mix of setup.py + pyproject.toml. Unfortunately just using pyproject.toml/build is still too broken with current distros, that's why I had to bring back setup.py.

I don't know what `gpep517` is, but isn't [PEP-517](https://peps.python.org/pep-0517/) about getting *rid* of setup.py? Why does this look at setup.py even?

Honestly this whole topic is still way too confusing, broken, and underdocumented. But I am very hesitant about adding hacks to classic setuptools/setup.py.

**Comment by martinpitt** [comment]
In other words: I don't consider current dbusmock as fit for building/releasing with PEP-517 tools. I only test/support setuptools/setup.py for now. For now I mostly use pyproject.toml as a central config file for mypy, pylint, etc. I'm happy to reattempt dropping setup.py in favor of pep517 in half a year or so, when this has matured a bit in distributions.

**Comment by mgorny** [comment]
> Right, I [noticed something similar](https://github.com/martinpitt/python-dbusmock/runs/7377001530?check_suite_focus=true) in CI yesterday. I considered this a regression in Debian unstable/Ubuntu devel series. setuptools did not need this `sys.path` wrangling for many years. I attributed this to the current mix of setup.py + pyproject.toml. Unfortunately just using pyproject.toml/build is still too broken with current distros, that's why I had to bring back setup.py.

It actually… depends on how you run Python. When you do:

    python foo.py

then the current directory is prepended to `sys.path`. Calling the script via shebang implies calling it via python, so the result is the same.

However, this whole thing is an inconsistent mess, and on top of it Python upstream is (finally!) realizing that adding the current directory on top of `sys.path` is a major security issue, so it might stop working like that in the future.

FWIU in a PEP517 builder you aren't supposed to rely on the current path being in `sys.path` (because you can't really know how it's going to be called).

> I don't know what `gpep517` is, but isn't [PEP-517](https://peps.python.org/pep-0517/) about getting _rid_ of setup.py? Why does this look at setup.py even?

It doesn't. However, all PEP517 tools have a backwards compatibility fallback to calling setuptools, and setuptools executes `setup.py`. Or to be more precise, it reads it and executes the code inside it rather than the script as an executable. That's how things work these days, so people don't have to implement two separate building modes for "old" and "new" packages.

> Honestly this whole topic is still way too confusing, broken, and underdocumented. But I am very hesitant about adding hacks to classic setuptools/setup.py.

It's not a hack. It's the correct solution. Relying on the current directory being on `sys.path` is a bad assumption.

**Comment by martinpitt** [comment]
> Relying on the current directory being on sys.path is a bad assumption.

But that's not my assumption, at least not in spirit. The assumption is that the directory which contains `setup.py` also contains the `dbusmock/` module, i.e. a relative import.  Perhaps this needs to be written in some other way, like `from . import dbusmock` or so, but I *don't* want to add `.` to sys.path. That is a workaround, it's not intended/required, and as you say this is dubious from a security POV.

**Comment by cschramm** [comment]
I just want to add that release 0.28.3 seems completely broken due to this. `pip install python-dbusmock==0.28.3` does not work as `pip wheel python-dbusmock==0.28.3` does not work due to the import in `setup.py`.

**Comment by martinpitt** [comment]
OK, meh -- I'll completely revert all the pyproject.toml changes then, until that is mature enough :cry:

**Comment by mgorny** [comment]
> > Relying on the current directory being on sys.path is a bad assumption.
> 
> But that's not my assumption, at least not in spirit. The assumption is that the directory which contains `setup.py` also contains the `dbusmock/` module, i.e. a relative import. Perhaps this needs to be written in some other way, like `from . import dbusmock` or so, but I _don't_ want to add `.` to sys.path. That is a workaround, it's not intended/required, and as you say this is dubious from a security POV.

Ok, here's another solution. If you don't mind not supporting ancient setuptools versions, the following `setup.cfg` should do the trick in a more modern (read: not-DYI) way:

```
[metadata]
version = attr: dbusmock.__version__
```

**Comment by martinpitt** [comment]
Ah, now I understand -- this is "just" fallout from commit 842d2506fb7184692e4907f8d5482c0da273ae48 , not from pyproject and such. But with adding `.` to sys.path this then fails further on:
```
❱❱❱ python3 -m build --sdist
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools >= 45, setuptools-scm)
* Getting dependencies for sdist...
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 363, in <module>
    main()
  File "/usr/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 345, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/usr/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 297, in get_requires_for_build_sdist
    return hook(config_settings)
  File "/tmp/build-env-6k9qhcos/lib64/python3.10/site-packages/setuptools/build_meta.py", line 181, in get_requires_for_build_sdist
    return self._get_build_requires(config_settings, requirements=[])
  File "/tmp/build-env-6k9qhcos/lib64/python3.10/site-packages/setuptools/build_meta.py", line 159, in _get_build_requires
    self.run_setup()
  File "/tmp/build-env-6k9qhcos/lib64/python3.10/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(code, locals())
  File "<string>", line 13, in <module>
  File "/var/home/martin/upstream/python-dbusmock/./dbusmock/__init__.py", line 15, in <module>
    from dbusmock.mockobject import (DBusMockObject, MOCK_IFACE,
  File "/var/home/martin/upstream/python-dbusmock/./dbusmock/mockobject.py", line 24, in <module>
    import dbus
ModuleNotFoundError: No module named 'dbus'
```

`dbus-python` is in setup.py's `install_requries` and pyproject.toml's `dependencies`, but apparently it does not install runtime dependencies for sdist. So importing `dbusmock` in setup.py is too much hassle.

@mgorny , thanks for the `[metadata]` trick! This works better indeed, tested with both `python3 -m build --sdist` and `python3 setup.py sdist`. I filed PR #144 with this.

**Comment by martinpitt** [comment]
I just [released 0.28.4](https://pypi.org/project/python-dbusmock/0.28.4/) with this fix and some more improvements, and `pip install` is happy again.
